File: Java/RemoveNthNodeFromEndOfList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
            
        ListNode ahead = head;
        ListNode behind = head;
        
        while(n-- > 0)
            ahead = ahead.next;
            
        if(ahead == null)
            return head.next;
            
        while(ahead.next != null){
            ahead = ahead.next;
            behind = behind.next;
        }
    
        behind.next = behind.next.next;
        return head;
    }
}

File: C++/RemoveNthNodeFromEndOfList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if(head ==NULL || head->next == NULL)
            return NULL;
        ListNode* p = head;
        ListNode* q = head;
        
        while(n--)
            p = p->next;
        if(!p)
            return head->next;
        while(p->next){
            p = p->next;
            q = q->next;
        }
        q->next = (q->next)->next;
        return head;
    }
};
